crates/provider/src/native.rs

```diff
@@ -2,531 +2,920 @@ use std::{
+    async fn logs(&self) -> Result<String, ProviderError> {
+        let inner = self.inner.read().await;
+        Ok(inner.filesystem.read_to_string(&inner.log_path).await?)
     }
 
-    async fn get_node_ip(&self, _node_name: &str) -> Result<IpAddr, ProviderError> {
-        Ok(LOCALHOST)
+    async fn dump_logs(&self, dest: PathBuf) -> Result<(), ProviderError> {
+        let logs = self.logs().await?;
+        Ok(self
+            .inner
+            .write()
+            .await
+            .filesystem
+            .write(dest, logs.as_bytes())
+            .await?)
     }
 
-    async fn get_port_mapping(&self, port: Port, node_name: &str) -> Result<Port, ProviderError> {
-        match self.process_map.get(node_name) {
-            Some(process) => match process.port_mapping.get(&port) {
-                Some(port) => Ok(*port),
-                None => Err(ProviderError::MissingNodeInfo(
-                    node_name.to_owned(),
-                    "port".into(),
-                )),
-            },
-            None => Err(ProviderError::MissingNodeInfo(
-                node_name.to_owned(),
-                "process".into(),
-            )),
-        }
+    async fn run_command(
+        &self,
+        _options: RunCommandOptions,
+    ) -> Result<ExecutionResult, ProviderError> {
+        todo!()
     }
-}
 
-#[cfg(test)]
-mod tests {
-    use std::{os::unix::process::ExitStatusExt, process::ExitStatus};
+    async fn run_script(
+        &self,
+        _options: RunScriptOptions,
+    ) -> Result<ExecutionResult, ProviderError> {
+        todo!()
+    }
 
-    use support::fs::mock::{MockError, MockFilesystem, Operation};
+    async fn copy_file_from_node(
+        &self,
+        _remote_src: PathBuf,
+        _local_dest: PathBuf,
+    ) -> Result<(), ProviderError> {
+        todo!()
+    }
 
-    use super::*;
+    async fn pause(&self) -> Result<(), ProviderError> {
+        let inner = self.inner.write().await;
+        let raw_pid = inner.process.id().unwrap();
+        let pid = Pid::from_raw(raw_pid.try_into().unwrap());
 
-    #[test]
-    fn new_native_provider() {
-        let native_provider: NativeProvider<MockFilesystem> =
-            NativeProvider::new("something", "/tmp", MockFilesystem::new());
+        kill(pid, Signal::SIGSTOP).unwrap();
 
-        assert_eq!(native_provider.namespace, "something");
-        assert_eq!(native_provider.tmp_dir, "/tmp");
-        assert_eq!(native_provider.command, "bash");
-        assert_eq!(native_provider.remote_dir, "/tmp/cfg");
-        assert_eq!(native_provider.data_dir, "/tmp/data");
+        Ok(())
     }
 
-    #[tokio::test]
-    async fn test_fielsystem_usage() {
-        let mut native_provider: NativeProvider<MockFilesystem> =
-            NativeProvider::new("something", "/tmp", MockFilesystem::new());
+    async fn resume(&self) -> Result<(), ProviderError> {
+        let inner = self.inner.write().await;
+        let raw_pid = inner.process.id().unwrap();
```

Comment:
we should not `unwrap` here.

Reply:
Good catch, still some work to do on error handling !